ucp/_libs/ucx_api.pyx

```diff
@@ -446,11 +446,88 @@ cdef class UCXListener(UCXObject):
         return int(<uintptr_t>self._handle)
 
 
+# Counter used as UCXRequest UIDs
+cdef int _ucx_request_counter = 0
+
+cdef class UCXRequest:
+    """Python wrapper of UCX request handle.
+
+    Notice, this class doesn't own the handle and multiple instances of
+    UCXRequest can point to the same underlying UCX handle.
+    Furthermore, UCX can modify/free the UCX handle without notice
+    thus we use `_uid` to make sure the handle hasn't been modified.
+    """
+    cdef:
+        ucp_request *_handle
+        int _uid
+
+    def __init__(self, uintptr_t req_as_int):
+        global _ucx_request_counter
+        cdef ucp_request *req = <ucp_request*>req_as_int
+        assert req != NULL
+        self._handle = req
+
+        cdef dict info = {"status": "pending"}
+        if self._handle.info == NULL:  # First time we are wrapping this UCX request
+            Py_INCREF(info)
+            self._handle.info = <PyObject*> info
+            _ucx_request_counter += 1
+            self._uid = _ucx_request_counter
```

Comment:
Would it be possible to use [`uuid.uuid4()`]( https://docs.python.org/3.8/library/uuid.html#uuid.uuid4 ) or perhaps [`secrets.randbits(32)`]( https://docs.python.org/3/library/secrets.html#secrets.randbits )? This would avoid needing a global variable.

Reply:
I must admit that I prefer a global counter in this case -- it is both faster and provide good debug info. But having said that, we could use `uuid.uuid4()` if avoiding global variables is important?